Misc Updates

Protonull opened the pull request.

Protonull committed:
AikarCommandManager :: mostly name update

Protonull committed:
CustomItems :: add static matcher

Protonull committed:
Merge branch 'master' of https://github.com/CivClassic/CivModCore into civ16

Protonull committed:
LocationAPI :: add getLocationWorld method

It removes the possibility for exceptions for unloaded or null worlds.

Protonull committed:
LocationAPI :: update areLocationsSameWorld

Now uses the new getLocationWorld method and allows for null world'd locations to match

Protonull committed:
Iteration :: add static size/length methods

Protonull committed:
Iteration :: add concatination shortcut

Protonull committed:
Iteration :: fix contains method bug

Protonull committed:
Iteration :: added array concatination tester

Protonull committed:
NullCoalescing :: fixed an equals bug

Protonull committed:
Iteration :: fix whitespace

Protonull committed:
LocationAPI :: mid block location + look at APIs

Protonull committed:
LocationIterator :: added location version of BlockIterator

This is a piece of code from the Bukkit forums. It allows you to do basic "block" iteration but it's done with locations instead so no block loading is done.

Protonull committed:
Add resets to CommandHandler and StandaloneCommandHandler

Protonull committed:
ACivMod :: add onUnload() and make some methods public

Protonull committed:
CustomItems :: remove the templated static matcher

It's an anti-pattern to match a custom item with criteria outside of the actual criteria as part of the custom item system itself.

Protonull committed:
MapUtils :: add Material parser

Protonull committed:
CustomItems :: refactor

Protonull committed:
POM :: bump version 1.8.1

Protonull committed:
CustomItem :: fix empty name bug

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/ACivMod.java:
```diff
@@ -23,7 +24,7 @@
 import vg.civcraft.mc.civmodcore.util.Iteration;
 
-public abstract class ACivMod extends JavaPlugin {
+public abstract class ACivMod extends JavaPlugin implements Closeable {
```
What's the point in this

Protonull replied:
It effectively provides an `onUnload()` method, though it's not as good as Java 9's [Cleanable](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.Cleanable.html), but we haven't officially moved from Java 8 yet. This is a pretty minor change that shouldn't effect people via changed signatures.

Maxopoly replied:
But there already is an onDisable? What does this add or how does this improve anything

Protonull replied:
Because `onDisable()` and `onUnload()` are two separate things. Just because the plugin has been disabled doesn't mean the instance no longer exists or has been abandoned. It allows a greater level of control without adding any complexity since no plugin is forced to use this.

Maxopoly replied:
Why would you ever disable a plugin except during startup and shutdown? That's bound to cause countless other problems and kill the server. We explicitly don't support any kind of runtime reload shenanigans (with some plugins providing their own exception).

Protonull replied:
Yes, and how is not supporting runtime reloading doing for you in terms of testing? Still having fun waiting for Banstick to disable?

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/util/LocationIterator.java:
```diff
@@ -0,0 +1,335 @@
+		this(entity, 0);
+	}
+
+	public Location getRelativeLocation(Location location, BlockFace face) {
```
https://hub.spigotmc.org/javadocs/spigot/org/bukkit/block/BlockFace.html#getDirection()

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/util/LocationIterator.java:
```diff
@@ -0,0 +1,335 @@
+	private BlockFace secondFace;
+	private BlockFace thirdFace;
+
+	public LocationIterator(World world, Vector start, Vector direction, double yOffset, int maxDistance) {
```
Could you explain what this entire class does that BlockIterator doesn't already?
This constructor is very big and hard to understand

Protonull replied:
> This is a piece of code from the Bukkit forums. It allows you to do basic "block" iteration but it's done with locations instead so no block loading is done.

From the commit message

Maxopoly replied:
So the only use case would be view ray tracing that's potentially further than render distance? Can't really think of a good use case for this, but ok why not.

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/util/Iteration.java:
```diff
@@ -1,241 +1,280 @@
+public final class Iteration {
+
+	@FunctionalInterface
+	public interface ElementAndBoolConsumer<T> {
```
What advantages does this provide over a BiConsumer ?
Seems to just add complexity

Protonull replied:
It avoids the potential null values of `Boolean`, which also why `IntConsumer` and `Predicate` exists.

Maxopoly replied:
Okay :man_shrugging:

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/util/Iteration.java:
```diff
@@ -1,241 +1,280 @@
+	 * @param latter The second of the two unknowns.
+	 * @return Returns either the former or the latter parameter if either match, or null if neither do.
+	 */
+	public static <T> T other(T base, T former, T latter) {
```
This and some of the ones below are unusable, because code containing them is unreadable. Just based on the method signature you have no idea what's going on

Protonull replied:
This one specifically can be removed, it was originally being used to detect which was the *other* chest in a double chest only using locations from the `InventoryHolder` thing. It hasn't been used elsewhere to my knowledge so can either be renamed to be clearer or outright removed.

Maxopoly replied:
`some` and `every` are the other main offenders

Protonull replied:
`some` and `every` where named after their Javascript counterparts, which I believe it stated explicitly in the their javadocs. We can rename them to `someMatch` and `allMatch` if necessary, but these are pretty old APIs by now.

Maxopoly commented on src/main/java/vg/civcraft/mc/civmodcore/custom/items/CustomItem.java:
```diff
@@ -0,0 +1,93 @@
+	public static CustomItem deserialize(final Map<String, Object> data) {
+		final NamespacedKey key = NamespaceAPI.fromString(castOrNull(String.class, data.get(ID_KEY)));
+		final String name = TextUtil.parse(castOrNull(String.class, data.get(NAME_KEY)));
+		final ItemCriteria[] extra = MapUtils.attemptGet(data, MapUtils::parseList, new ArrayList<>(), EXTRA_KEY)
```
No clue what this does without looking up those MapUtils methods, which makes it unreadable code imo

Protonull replied:
With all due respect, that exact same argument could be made towards your `CoreConfigManager.parseList(config, key, parser);`

Maxopoly replied:
I feel like the second arg in `parseList` is non obvious, but otherwise eh (though I'm obviously very biased regarding it). That being said I don't see how that defends what you're doing here in any way.

Protonull replied:
The function signature is `MapUtils.attemptGet(map, parser, fallback, key);` the reason why the key is last is because it's a rest spread, which allows for this:
```java
long connectionTimeout = MapUtils.attemptGet(data, NumberConversions::toLong, 10_000L,
		"connectionTimeout", "connection_timeout");
```

Maxopoly commented:
Generally not very convinced of these methods which do nothing but adding a null check, but make the code harder to read because you need to know what it does.

Protonull committed:
POM :: swap out spigot for paper

Maxopoly commented on pom.xml:
```diff
@@ -30,8 +30,8 @@
-      <groupId>org.spigotmc</groupId>
-      <artifactId>spigot</artifactId>
+      <groupId>com.destroystokyo.paper</groupId>
+      <artifactId>paper</artifactId>
```
Does this make paper a hard requirement for building? paper-api is already inherited from style-guide/parent, no?

Protonull replied:
It would, and no, neither paper nor spigot were added as dependencies within the parent pom, though the repositories were added for easy dependency addition.

Maxopoly replied:
> though the repositories were added for easy dependency addition.

Ah yeah right

> It would

I don't think that's desireable then. It makes Paper a hard requirement for the entire civ stack.

Protonull replied:
Without it, we wont be able to use Paper's de-obfuscation methods or any custom features

Maxopoly replied:
Do we _need_ any of  those? As far as this PR goes it makes things a bit nicer and NMS a bit rarer, but that's it

Protonull replied:
It *is* more of a want than an explicit need, but as you said the other day, there's no reason not to use it by now :P

Protonull replied:
Reverted the Paper commits

Protonull committed:
NBTCompound :: updated UUIDs methods to use Paper APIs

Protonull committed:
Post Paper fixes

Protonull committed:
Iteration :: add removeLastElement method

Protonull committed:
Iteration :: make removeLastElement return the removed element

Protonull committed:
NBTSerialization :: replace final requirement

This was added when ItemExchange was being developed to prevent anonymous classes. Little did I know that there was a far easier way of doing that. Tuts in my direction please.

Protonull committed:
Iteration :: fix some of the complaints from the PR

Protonull committed:
TextUtil :: add isNullOrEmpty check for BaseComponent

Protonull committed:
Iteration :: hidden bug fixed

I've deprecated but also changed the signature of `Iteration.isNullOrEmpty(T... array)` to `Iteration.isNullOrEmpty(T[] array)` which may cause problems, but not fixing this would cause more problems down the line due to hidden bugs. Basically, for some things, such as Maps, you may forget (as I did) that they aren't collections, but will still look correct to the compiler because it's using the generic array version instead of the collection version, which bypasses the actual check you're doing about whether the Map is null or empty, which for that I've created a specific check in MapUtils. I've deprecated the generic array version because a better version exists thanks to `ArrayUtils`.

Protonull committed:
UuidUtils :: new set of APIs for UUIDs

Protonull committed:
ACivMod :: Max doesn't want to support runtime reloading in any regards

Protonull committed:
Revert "POM :: swap out spigot for paper"

This reverts commit 54d67aa33b2661aac8ab508db259362b4415d023.

Protonull committed:
Revert "Post Paper fixes"

This reverts commit e7deee51872212b6fa140aa7876627a4cc0b15fb.

Protonull committed:
Title :: keep this

Protonull committed:
POM :: change Jenkins back to DevotedMC

Protonull committed:
NBTCompound :: fix UUID thing from Paper revert

Protonull committed:
Merge branch 'master' of https://github.com/CivClassic/CivModCore into main

Protonull committed:
NBTCompound :: add UUID array methods

UUIDs are the only explicit type that doesn't have an array getter and setter.

Protonull committed:
Iteration :: addAll API

Protonull committed:
ACivMod :: remove unused import

Protonull committed:
NBTSerialization :: add search API

Protonull committed:
ACivMod :: added saveDefaultResourceAs API

Protonull committed:
MapUtils :: add getKeyFromValue API

Protonull committed:
TextUtil :: add QoL constructor for TextComponent

Protonull commented:
Closing this PR because the additions are too numerous and broad in scope
